File: src/main/java/net/sf/rails/game/state/HashMultimapState.java
package net.sf.rails.game.state;

import com.google.common.collect.HashMultimap;
import com.google.common.collect.Multimap;

/**
 * A stateful version of a HashMultimap
 */

public final class HashMultimapState<K,V> extends MultimapState<K,V> {
    
    private final HashMultimap<K,V> map  = HashMultimap.create();
    
    private HashMultimapState(Item parent, String id) {
        super(parent, id);
    }
    
    /** 
     * Creates an empty HashMultimapState 
     */
    public static <K,V> HashMultimapState<K,V> create(Item parent, String id){
        return new HashMultimapState<K,V>(parent ,id);
    }
    
    @Override
    protected Multimap<K,V> getMap() {
        return map;
    }
    
}

File: dotnet/RailsLib.Net/Net/Game/State/HashMultimapState.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using Wintellect.PowerCollections;

namespace GameLib.Net.Game.State
{
    public class HashMultimapState<K, V> : MultimapState<K, V>
    {
        private readonly Wintellect.PowerCollections.MultiDictionary<K, V> map;

        private HashMultimapState(IItem parent, string id) : base(parent, id)
        {
            map = new Wintellect.PowerCollections.MultiDictionary<K, V>(true);
        }

        /** 
         * Creates an empty HashMultimapState 
         */
        public static HashMultimapState<K, V> Create(IItem parent, String id)
        {
            return new HashMultimapState<K, V>(parent, id);
        }

        protected override MultiDictionaryBase<K, V> Clone()
        {
            return map.Clone();
        }

        protected override MultiDictionaryBase<K, V> GetMap()
        {
            return map;
        }
    }
}
